### ISSUE TITLE
Integral regression result using softmax will be affected by the location p where the heatmap value is 0.

### ISSUE THREAD

**Issue by r1c7** [JimmySuen/integral-human-pose#8]
We only consider one keypoint.
If the heatmap is
[[0.  , 0.  , 0.  , 0.  , 0.  , 0.  , 0.  ],
 [0.  , 0.  , 0.  , 0.  , 0.  , 0.  , 0.  ],
 [0.  , 0.  , 0.  , 0.  , 0.  , 0.  , 0.  ],
 [0.  , 0.1 , 0.  , 0.  , 0.  , 0.  , 0.  ],
 [0.2 , 0.3 , 0.2 , 0.2 , 0.  , 0.  , 0.  ],
 [0.1 , 0.4 , 0.05, 0.  , 0.  , 0.  , 0.  ],
 [0.2 , 0.2 , 0.2 , 0.2 , 0.  , 0.  , 0.  ]]

After using the softmax, the normalized heatmap become
[[0.01935992, 0.01935992, 0.01935992, 0.01935992, 0.01935992, 0.01935992, 0.01935992],
 [0.01935992, 0.01935992, 0.01935992, 0.01935992, 0.01935992, 0.01935992, 0.01935992],
 [0.01935992, 0.01935992, 0.01935992, 0.01935992, 0.01935992, 0.01935992, 0.01935992],
 [0.01935992, 0.02139602, 0.01935992, 0.01935992, 0.01935992, 0.01935992, 0.01935992],
 [0.02364626, 0.02613315, 0.02364626, 0.02364626, 0.01935992, 0.01935992, 0.01935992],
 [0.02139602, 0.0288816 , 0.02035252, 0.01935992, 0.01935992, 0.01935992, 0.01935992],
 [0.02364626, 0.02364626, 0.02364626, 0.02364626, 0.01935992, 0.01935992, 0.01935992]]  .

So integral regression result will be affected by the  value 0.01935992 which is calculated using softmax.

**Comment by lck1201** [comment]
When using integral, only after softmax, value in heatmap stands for possibility, before that, the value doesn't have any semantics.  In fact, if you print the direct output of network, there are many negatives, but it doesn't matter. @crsailing

**Comment by r1c7** [comment]
If we use heatmap of groundtruth to test, using the integral after softmax, we cannot get the accurate label position.

**Comment by strawberryfg** [comment]
> If we use heatmap of groundtruth to test, using the integral after softmax, we cannot get the accurate label position.

Try scaling the ground truth heatmap by a factor of say, 30.0 or 50.0, and then softmax $\rightarrow$ integral. Plot the function of softmax in google, and you will find this sketchy trick.

**Comment by lck1201** [comment]
@crsailing  you need to normalize the heatmap that the summed value equals to one

**Comment by lck1201** [comment]
@crsailing  The GT heatmap you assumed, in fact, is totally not the same with prediction of the network. Try to print the direct output of network, and see what the heatmap looks like. By the way, the integral is a new method, not related to onehot/gaussian heatmap, so from this perspective, you cannot assume a GT heatmap, applying integral on it, and expect getting accurate position. It doesn't make any sense.

**Comment by chaneyddtt** [comment]
Hi, 
if the output is not like a gaussian heatmap, how do you train the network with both heatmap and intergral loss, namely I1 in your paper. The output should be something like a gaussian heatmap since you supervise the network with the gaussian heatmap before the intergral operation?

**Comment by lck1201** [comment]
@chaneyddtt  Hi, in practice, we train the network using gaussian loss (H1) and finetune with integral loss(l1)
> For integral regression methods (I1, I2, I3, and their multi-stage versions), the network is pre-trained only using heat map loss (thus their H versions) and then, only integral loss is used. We found this training strategy working slightly better than training from scratch using both losses.

**Comment by chaneyddtt** [comment]
Ok, thanks for clarifying.

**Comment by Ericjiejie** [comment]
@lck1201 Hi, is the heatmap on the right column of Figure 2 in the your paper(https://arxiv.org/pdf/1711.08229v1.pdf) ouput from network or softmax?

**Comment by lck1201** [comment]
@Ericjiejie Hi, please email @JimmySuen . I am not quite sure.

**Comment by Ericjiejie** [comment]
Ok, thanks .But I don't know if he will return my email.
The reason why I ask is because you say that the output of the network is  different from that of GT Heatmap(you talk to @r1ch88), so I think the heatmap on Figure 2 comes from softmax.

**Comment by lck1201** [comment]
I guess it's the output of softmax, beacause only the value after softmax represents possibility.

**Comment by Ericjiejie** [comment]
I agree with you.
I have another question. If the output of the network is not like Gaussian heatmap, even very different, why does the author use GT Heatmap for supervision? Will adding this supervision not affect network training?

**Comment by Ericjiejie** [comment]
@lck1201 Hi, where is the I1 loss(heatmap loss and joint loss) in your code? I want to know the whole process of the integral version(I1), because I have some problems to look for in the code.In detail, I don’t understand the heatmap supervision part and want to know what the output of the network is.

**Comment by lck1201** [comment]
check https://github.com/JimmySuen/integral-human-pose/blob/master/pytorch_projects/common_pytorch/common_loss/integral.py

**Comment by Lixiuxiu2020** [comment]
I checked it, but there is no heatmap loss, only joint loss.
